File: leetcode_412.cpp
class Solution {
public:
  int longestValidParentheses(string s) {
    int max_len = 0, last = -1;
    stack<int> lefts;

    for (auto i = 0; i < s.size(); i++) {
      if (s[i] == '(') {
        lefts.push(s[i]);
      } else {
        if (lefts.empty()) {
          last = i;
        } else {
          lefts.pop();
          if (lefts.empty()) {
            max_len = max(max_len, i - last);
          } else {
            max_len = max(max_len, i - lefts.top());
          }
        }
      }
    }

    return max_len;
  }
}

File: leetcode_412.py
class Solution:
    # @param s, a string
    # @return an integer
    def longestValidParentheses(self, s):
        lefts = []
        maxLen = 0
        last = -1
        for i in range(len(s)):
            c = s[i]
            if c == '(':
                lefts.append(i)
            elif c == ')':
                if len(lefts) == 0:
                    last = i
                else:
                    lefts.pop()
                    if len(lefts) == 0:
                        maxLen = max(maxLen, i - last)
                    else:
                        maxLen = max(maxLen, i - lefts[len(lefts) - 1])
        return maxLen

s = Solution()
print s.longestValidParentheses('()())')
print s.longestValidParentheses('()(()')
print s.longestValidParentheses('()(())')
